File: app/src/main/java/tdc/com/bidirectionaldijkstrasalgorithm/model/Graph.java
package tdc.com.bidirectionaldijkstrasalgorithm.model;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;

class Graph {

  private Map<String, List<Edge>> graphMap;

  Graph() {
    graphMap = new HashMap<>();
  }

  Graph(int n) {
    graphMap = new HashMap<>(n);
  }

  List<Edge> getEdges(String node) {
    return graphMap.get(node);
  }

  void addDirectedEdge(String from, String to, double distance) {
    graphMap.computeIfAbsent(from, x -> new LinkedList<>()).add(new Edge(to, distance));
  }

  void addUndirectedEdge(String from, String to, double distance) {
    addDirectedEdge(from, to, distance);
    addDirectedEdge(to, from, distance);
  }

  public Graph reverseGraph() {
    Graph reverseGraph = new Graph(graphMap.size());
    for (String from : graphMap.keySet()) {
      for (Edge edge : graphMap.get(from)) {
        reverseGraph.addDirectedEdge(edge.to, from, edge.distance);
      }
    }
    return reverseGraph;
  }

  public int size() {
    return graphMap.size();
  }

  public Set<String> getNodeSet() {
    return graphMap.keySet();
  }


}

File: app/src/main/java/tdc/com/bidirectionaldijkstrasalgorithm/model/forkotlin/Graph.kt
package tdc.com.bidirectionaldijkstrasalgorithm.model.forkotlin

import java.util.*

data class Graph(val n: Int) {
    private val graphMap: MutableMap<String?, MutableList<Edge>> = HashMap(n)

    fun getEdges(node: String): List<Edge>? {
        return graphMap[node]
    }

    fun addDirectedEdge(from: String, to: String, distance: Double) {
        graphMap.computeIfAbsent(from) { LinkedList() }.add(Edge(to, distance))
    }

    fun addUndirectedEdge(from: String, to: String, distance: Double) {
        addDirectedEdge(from, to, distance)
        addDirectedEdge(to, from, distance)
    }

    fun reverseGraph(): Graph {
        val reverseGraph = Graph(graphMap.size)
        for (from in graphMap.keys) {
            for (edge in graphMap[from]!!) {
                reverseGraph.addDirectedEdge(edge.to, from!!, edge.distance)
            }
        }
        return reverseGraph
    }

    fun size(): Int {
        return graphMap.size
    }

    val nodeSet: MutableSet<String?>
        get() = graphMap.keys
}